Classes/New Issue/GithubClient+Template.swift

```diff
@@ -0,0 +1,115 @@
+//  Created by Ehud Adler on 11/11/18.
+//  Copyright © 2018 Ryan Nystrom. All rights reserved.
+//
+
+import Foundation
+import GitHubAPI
+import GitHubSession
+import Squawk
+
+private let githubIssueURL = ".github/ISSUE_TEMPLATE"
+let tempalateRegex = ".github/ISSUE_TEMPLATE"
+
+struct IssueTemplateRepoDetails {
+    let owner: String
+    let name: String
+    let defaultBranch: String
+}
+
+extension GithubClient {
+
+    private func fetchTemplateFile(
+        repoDetails: IssueTemplateRepoDetails,
+        filename: String,
+        testingFile: String? = nil,
+        completion: @escaping (Result<String>) -> Void
+        ) {
+
+        // For Testing.
+        if let testingFile = testingFile { completion(.success(testingFile)) }
+
+        fetchFile(
+            owner: repoDetails.owner,
+            repo: repoDetails.name,
+            branch: repoDetails.defaultBranch,
+            path: "\(githubIssueURL)/\(filename)") { result in
+                switch result {
+                case .success(let file): completion(.success(file))
+                case .error(let error): completion(.error(error))
+                case .nonUTF8: completion(.error(nil))
+                }
+        }
+    }
+
+    func createTemplate(
+        repoDetails: IssueTemplateRepoDetails,
+        filename: String,
+        testingFile: String? = nil,
+        completion: @escaping (Result<IssueTemplate>) -> Void
+        ) {
+
+        fetchTemplateFile(repoDetails: repoDetails, filename: filename, testingFile: testingFile) { result in
+            switch result {
+            case .success(let file):
+                let nameAndDescription = IssueTemplateHelper.getNameAndDescription(fromTemplatefile: file)
+                if let name = nameAndDescription.name {
+                    let cleanedFile = IssueTemplateHelper.cleanText(file: file)
+                    completion(.success(IssueTemplate(title: name, template: cleanedFile)))
+                } else {
+                    completion(.error(nil))
+                }
+            case .error(let error):
+                completion(.error(error))
+            }
+        }
+    }
+
+    func createNewIssue(
+        repoDetails: IssueTemplateRepoDetails,
+        completion: @escaping (Result<[IssueTemplate]>) -> Void
+        ) {
+
+        var templates: [IssueTemplate] = []
+
+        // Create group.
+        // We need this since we will be making multiple async calls inside a for-loop.
+        let templateGroup = DispatchGroup()
+
+        fetchFiles(
+            owner: repoDetails.owner,
+            repo: repoDetails.name,
+            branch: repoDetails.defaultBranch,
+            path: githubIssueURL) { result in
+                switch result {
+                case .success(let files):
+                    for file in files {
+                        templateGroup.enter()
+                        self.createTemplate(repoDetails: repoDetails, filename: file.name, completion: { result in
```

Comment:
I'm going to block this PR on the multiple-requests part. It's just too wasteful IMO as-is. Some options:

1. Do a fetch for the directory and just show the file names in the list instead of the template title. Then fetch the contents of the template once its been selected.
2. Only fetch this info once then cache it. Always use the cache but allow the user to pull-to-refresh the templates list.

I favor option 1 personally.

Reply:
Sorry to jump in on here @rnystrom, and I know it's a massive shift in direction but could I suggest a third option here?

If we use the v4 API (ooh bois he's said the danger word) we can actually do this with a single request which returns the contents of every template within the directory?

This request: (Notably, we would need to replace the default branch and repo owner/name)
```gql
query {
  repository(owner: "GitHawkApp", name:"GitHawk") {
    object(expression: "master:.github/ISSUE_TEMPLATE") {
      ... on Tree {
        entries {
          name
          object {
            ... on Blob {
              text
            }
          }
        }
      }
    }
  }
}
```

Returns us this response:
```json
{
  "data": {
    "repository": {
      "object": {
        "entries": [
          {
            "name": "bug_report.md",
            "object": {
              "text": "---\nname: Bug report  (etc etc, I've truncated this)"
            }
          },
          {
            "name": "feature_request.md",
            "object": {
              "text": "---\nname: Feature request (etc etc, I've truncated this)"
            }
          }
        ]
      }
    }
  }
}
```

One request, all the information? Opinion Ryan?

Reply:
Lovd it!